scikeras/utils/__init__.py

```diff
@@ -0,0 +1,109 @@
+from inspect import isclass
+from typing import Callable, Union
+
+import numpy as np
+
+from tensorflow.keras.losses import Loss
+from tensorflow.keras.losses import get as get_loss
+from tensorflow.keras.losses import serialize as serialize_loss
+from tensorflow.keras.metrics import Metric
+from tensorflow.keras.metrics import get as get_metric
+from tensorflow.keras.metrics import serialize as serialize_metric
+
+
+def loss_name(loss: Union[str, Loss, Callable]) -> str:
+    """Retrieves a loss's full name (eg: "mean_squared_error").
+
+    Parameters
+    ----------
+    loss : Union[str, Loss, Callable]
+        Instance of Keras Loss, loss callable or string
+        shorthand (eg: "mse") or full name ("mean_squared_error").
+
+    Returns
+    -------
+    str
+        [description]
+
+    Examples
+    --------
+    >>> loss_name("BinaryCrossentropy")
+    'BinaryCrossentropy'
+    >>> loss_name("binary_crossentropy")
+    'binary_crossentropy'
+    >>> import tensorflow.keras.losses as losses
+    >>> loss_name(losses.BinaryCrossentropy)
+    'BinaryCrossentropy'
+    >>> loss_name(losses.binary_crossentropy)
+    'binary_crossentropy'
+
+    Raises
+    ------
+    ValueError
+        In case of an unknown loss.
+    """
+    if isclass(loss):
+        try:
+            loss = loss()  # get_loss accepts instances, not classes
+        except TypeError:
+            # convert TypeError -> ValueError for consistency with errors
+            # raised by get_loss
+            raise ValueError(f"Unknown loss: {loss}")
+    try:
+        loss = serialize_loss(get_loss(loss))
+    except ValueError:
+        # Error messages are different for metrics or losses; homogenize them
+        raise ValueError(f"Unknown loss: {loss}")
```

Comment:
Why is this try/except necessary? A ValueError is already raised with a clear error message.

I see that "error messages are different for metrics and losses," but I'm not detecting a huge difference in error messages from Tensorflow's error:

* "ValueError: Unknown loss function: mse2"
* "ValueError: Unknown metric function: acc2"

<details>

``` python
[ins] In [7]: loss = "mse2"

[ins] In [8]: serialize_loss(get_loss(loss))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-1339b37868fa> in <module>
----> 1 serialize_loss(get_loss(loss))

~/anaconda3/envs/dask-ml-test/lib/python3.8/site-packages/tensorflow/python/keras/losses.py in get(identifier)
   1893   if isinstance(identifier, six.string_types):
   1894     identifier = str(identifier)
-> 1895     return deserialize(identifier)
   1896   if isinstance(identifier, dict):
   1897     return deserialize(identifier)

~/anaconda3/envs/dask-ml-test/lib/python3.8/site-packages/tensorflow/python/keras/losses.py in deserialize(name, custom_objects)
   1848       A Keras `Loss` instance or a loss function.
   1849   """
-> 1850   return deserialize_keras_object(
   1851       name,
   1852       module_objects=globals(),

~/anaconda3/envs/dask-ml-test/lib/python3.8/site-packages/tensorflow/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    375       obj = module_objects.get(object_name)
    376       if obj is None:
--> 377         raise ValueError(
    378             'Unknown ' + printable_module_name + ': ' + object_name)
    379     # Classes passed by name are instantiated with no args, functions are

ValueError: Unknown loss function: mse2
```

</details>

<details>

``` python
[ins] In [10]: metric = "acc2"

[ins] In [11]: serialize_metric(get_metric(metric))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-38a06ad5e0be> in <module>
----> 1 serialize_metric(get_metric(metric))

~/anaconda3/envs/dask-ml-test/lib/python3.8/site-packages/tensorflow/python/keras/metrics.py in get(identifier)
   3481     return deserialize(identifier)
   3482   elif isinstance(identifier, six.string_types):
-> 3483     return deserialize(str(identifier))
   3484   elif callable(identifier):
   3485     return identifier

~/anaconda3/envs/dask-ml-test/lib/python3.8/site-packages/tensorflow/python/keras/metrics.py in deserialize(config, custom_objects)
   3437       A Keras `Metric` instance or a metric function.
   3438   """
-> 3439   return deserialize_keras_object(
   3440       config,
   3441       module_objects=globals(),

~/anaconda3/envs/dask-ml-test/lib/python3.8/site-packages/tensorflow/python/keras/utils/generic_utils.py in deserialize_keras_object(identifier, module_objects, custom_objects, printable_module_name)
    375       obj = module_objects.get(object_name)
    376       if obj is None:
--> 377         raise ValueError(
    378             'Unknown ' + printable_module_name + ': ' + object_name)
    379     # Classes passed by name are instantiated with no args, functions are

ValueError: Unknown metric function: acc2
```

</details>

Reply:
I'm not 100% sure, but I know that at least in TF 2.2.0, there is a typo and you get `Unknown loss function:mse2` (i.e. missing a space). I also find it confusing that it always has the word `function` in there, even if it's a class. But I agree with you that's not much of a difference. Maybe I was testing with some random TF version where there was more differences?

Reply:
Here's another thing I was trying to catch:
```python3
class Test:
    pass
losses.serialize(losses.get(Test()))  # ValueError: ('Could not interpret loss function identifier...
```
Like I said, I ideally think these methods should have a single failure mode.

Reply:
Another couple failure modes:
```python3
metrics.serialize(metrics.get(dict()))  # ValueError: Improper config format
metrics.serialize(metrics.get(list()))  # ValueError: Could not interpret metric function
losses.serialize(losses.get(dict()))  # ValueError: Improper config format
losses.serialize(losses.get(list()))  # ValueError: ('Could not interpret loss function
```

Reply:
We now handle `metrics.serialize(metrics.get(dict()))` internally as a type error. So we are left with the question: do we want to catch the `ValueError` to homogenize `Could not interpret metric function`, `('Could not interpret loss function`, `Unknown loss function: mse2` and `Unknown loss function:mse2` into a single error message?

Reply:
Resolved in #102